datafusion/functions-nested/src/array_has.rs

```diff
@@ -121,6 +123,43 @@ impl ScalarUDFImpl for ArrayHas {
         Ok(DataType::Boolean)
     }
 
+    fn simplify(
+        &self,
+        mut args: Vec<Expr>,
+        _info: &dyn datafusion_expr::simplify::SimplifyInfo,
+    ) -> Result<ExprSimplifyResult> {
+        let [haystack, _needle] = take_function_args(self.name(), &args)?;
+
+        // if the haystack is a constant list, we can use an inlist expression which is more
+        // efficient because the haystack is not varying per-row
+        if let Expr::Literal(ScalarValue::List(array)) = haystack {
+            assert_eq!(array.len(), 1); // guarantee of ScalarValue
+            if let Ok(scalar_values) =
+                ScalarValue::convert_array_to_scalar_vec(array.as_ref())
+            {
+                assert_eq!(scalar_values.len(), 1);
+                let list = scalar_values
+                    .into_iter()
+                    .flatten()
+                    .map(Expr::Literal)
+                    .collect();
+
+                // ok to pop here, we will not use args again
+                let needle = args.pop().unwrap();
+                return Ok(ExprSimplifyResult::Simplified(Expr::InList(InList {
+                    expr: Box::new(needle),
+                    list,
+                    negated: false,
+                })));
+            }
+        }
+
+        // TODO: support LargeList / FixedSizeList?
+        // (not supported by `convert_array_to_scalar_vec`)
```

Comment:
Input wanted on whether this should be resolved (maybe by handling those data types in `convert_aray_to_scalar_vec`).

Reply:
I suggest doing the work in a follow on ticket / PR. So that would mean:
1. Filing a ticket with a reproducer (perhaps modified from https://github.com/apache/datafusion/issues/14533)
2. leaving a TODO with a link to that ticket here

Reply:
Filed as #15389